### ISSUE TITLE
backward pass with manually specified gradient

### ISSUE THREAD

**Issue by zeroexcuses** [LaurentMazare/ocaml-torch#17]
https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#sphx-glr-beginner-blitz-autograd-tutorial-py has an example showing:

```
x = torch.randn(3, requires_grad=True)

y = x * 2
while y.data.norm() < 1000:
    y = y * 2

v = torch.tensor([0.1, 1.0, 0.0001], dtype=torch.float)
y.backward(v)
```

The key line here is `y.backward(v)`

The OCaml bindings for `backward` only take a single tensor:

```
grep "val backward" . -R
./torch/optimizer.mli:val backward_step : ?clip_grad:Clip_grad.t -> t -> loss:Tensor.t -> unit
./wrapper/wrapper.mli:  val backward : ?keep_graph:bool -> ?create_graph:bool -> t -> unit
```

If we search for "t -> t -> unit", nothing useful shows up:

```
rep "t -> t -> unit" . -R
./torch/tensor.mli:val ( += ) : t -> t -> unit
./torch/tensor.mli:val ( -= ) : t -> t -> unit
./torch/tensor.mli:val ( *= ) : t -> t -> unit
./torch/tensor.mli:val ( /= ) : t -> t -> unit
./torch/optimizer.mli:val step : ?clip_grad:Clip_grad.t -> t -> unit

```

What is the OCaml way of running a backward pass with a custom gradient ?

**Comment by LaurentMazare** [comment]
I think pointwise multiplication should cover this use case but maybe I'm missing something ?
```ocaml
let x = Tensor.randn [3] ~requires_grad:true;;
let rec loop x =
  if Tensor.norm x |> Tensor.to_float0_exn < 1000.
  then loop Tensor.(x * f 2.)
  else x;;
let y = loop x;;
Tensor.backward y;;
Tensor.(grad x * of_float1 [| 0.1; 1.0; 1e-4|]) |> Tensor.print;;
```
(note that because of `randn` the value you get may be different from one run to the other)

**Comment by zeroexcuses** [comment]
In this particular case, yes -- but we can't do this in general.

Let Y = var we want to run backward pass from.
Let X = input vars.

In the general case, there's no guarantee that X.size() == Y.size().

=====

I'm even more confused now. y is a tensor of shape [3]. y is NOT a scalar.

How does "Tensor.backward y" even work? Is it assuming that grad)_y = [1.0, 1.0, 1.0] ?

**Comment by zeroexcuses** [comment]
So here's my understanding, numbered so you can point out which one is wrong:

  (1) Let X be a Tensor w/ precisely 1 element. If we run a backward pass from X, we just do it assuming X_grad = 1.0 // perfectly reasonable, as X is generally the "error" term.

  (2) Let X be a Tensor w/ > 1 elements. If we run a backward pass from X, we need a starting point grad_X which says "for each element of X[i] of X, if we change X[i], how much do we change final error term"

  (3) Therefore, how does "Tensor.backward y" when y.size() = [3] even work? Are we assuming grad_y = [1.0, 1.0, 1.0] ?

**Comment by LaurentMazare** [comment]
I'm not sure to agree that X.size() != Y.size. If you run your pytorch example with ` y.backward(torch.tensor([0.1, 1.0, 0.0001, 1.]))` you will get an error because the shape are different.
I don't think this compute the jacobian which would be the general case where x and y can have different sizes, but instead this computes the vector-jacobian product.

**Comment by LaurentMazare** [comment]
(also I don't thin kany of your point is wrong)

**Comment by zeroexcuses** [comment]
I did not frame my question properly. Let's reset and consider a concrete example:

```
let dtype = Tchc.Kind.Float in
let device = Tchc.Device.Cpu in
let m = Ten.randn [5; 2] ~device:device ~kind:dtype in (* this is const matrix *)
let x = Ten.randn [2] ~device:device ~kind:dtype ~requires_grad:true in
let y = Ten.mm m x in
let _ = Ten.backward y in
(Ten.size m, Ten.size x, Ten.size y, Ten.grad x);;

```

I got output:

```
- : int list * int list * int list * Ten.t = ([5; 2], [2], [5], 
-2.5771
 0.6864
[ Variable[CPUType]{2} ]
)

```

So x has dim [2], y has diim [5], y_grad has dim [5].

We now have a situation where x.size() != y.size(). This means:

(1) we don't do point-wise mul of x_grad and y_grad
(2) what does `Ten.backward y` even do? Did it assume that y_grad = [1; 1.; 1.; 1.; 1.] ?

**Comment by LaurentMazare** [comment]
(1) Actually I think we should still do pointwise multiplication but I was wrong applying it to `grad x`, we should more do `Tensor.(backward (y * ...))`.
(2) I think so, note that it's actually handled by pytorch and not the ocaml-torch layer so I'm not actually sure as I haven't looked at the code.

**Comment by zeroexcuses** [comment]
You're right, doing backward path on pointwise mul of y, y_grad works mathematically.

Thanks for your infinite patience in answering all these low level questions!
